Add LibAssetData to contracts/asset-proxy

feuGeneA opened the pull request:
## Description

[Asana task](https://app.asana.com/0/711604532004534/1116922335701244/f)

-   [x] Encoding asset data
-   [x] Decoding asset data
-   [x] Querying balances
-   [x] Querying allowances (all done except for multi-asset)
-   [x] Batch combinations of the above
-   [x] Documentation
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

## Testing instructions

Tests are integrated with `yarn test` and therefore CircleCI.

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->

<!-- * Bug fix (non-breaking change which fixes an issue) -->

* New feature (non-breaking change which adds functionality)

feuGeneA commented on contracts/asset-proxy/test/lib_asset_data.ts:
```diff
@@ -0,0 +1,309 @@
+    address: '0x1dc4c1cefef38a777b15aa20260a54e584b16c48',
+    assetData: '0xf47261b00000000000000000000000001dc4c1cefef38a777b15aa20260a54e584b16c48',
+};
+const KNOWN_ERC721_ENCODING = {
```
Note that these known good encodings were copied from ../../../../packages/order-utils/test.

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,220 @@
+    function allowance(address owner, address spender, bytes memory assetData)
+        // should we rename to getApproval()? "Allowance" seems out of vogue.
+        public
+        view
```
See comment above.

feuGeneA commented on contracts/asset-proxy/test/lib_asset_data.ts:
```diff
@@ -0,0 +1,309 @@
+        // Somewhat re-inventing the wheel here, but the prior art currently
+        // exists only as an unexported test util in the erc1155 package
+        // (Erc1155Wrapper.mintFungibleTokensAsync() in erc1155/test/utils/).
+        // This is concise enough to justify duplication, but it sure is ugly.
```
See comment.  That concise but ugly code is:
```
erc1155TokenId = ((await new LogDecoder(web3Wrapper, erc1155Artifacts).getTxWithDecodedLogsAsync(
    await erc1155MintableContract.create.sendTransactionAsync('uri:Dummy', /*isNonFungible:*/ false),
)).logs[0] as LogWithDecodedArgs<ERC1155TransferSingleEventArgs>).args.id;
```

feuGeneA commented on contracts/utils/test/lib_address_array.ts:
```diff
@@ -23,22 +23,14 @@ describe('LibAddressArray', () => {
-    after(async () => {
-        await blockchainLifecycle.revertAsync();
-    });
-    before(async () => {
```
I cloned this test to start my own, and noticed this inefficiency in there, and figured I'd leave it better off than I found it.

abandeali1 commented:
I would also prefer to standardize all of the function names to `getBalance`, `getAllowance`, etc. Finally, we should also add `getBalanceAndAllowance` and `getBalancesAndAllowances` functions in addition to the existing batch functions.

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,228 @@
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/LibBytes.sol";
+import "@0x/contracts-erc1155/contracts/src/interfaces/IERC1155Mintable.sol";
```
Ignorance. :)  I guess because I wrote the tests using a `Mintable` I then wrote the implementation using the same interface.  TDD for the lose! :)  Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/f2055d215da9670e0d5a061e7cbaeee22ebddd5d .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,228 @@
+            );
+            address[] memory owners = new address[](1);
+            owners[0] = owner;
+            return IERC1155Mintable(tokenAddress).balanceOfBatch(owners, tokenIds)[0];
```
Reading the spec more closely, I think your interpretation is right.  Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/c890621e3ef967a4d9cf89f700d6eaec13267e86 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+        return abi.encodeWithSelector(ERC20_PROXY_ID, tokenAddress);
+    }
+
+    function decodeERC20AssetData(bytes memory assetData)
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/c4ec07e2883f787345c5eadca291395e399e9fa3 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+            // solhint-disable-next-line indent
+            (, bytes[] memory nestedAssetData) = abi.decode(assetDataBody, (uint256[], bytes[]));
+            for (uint256 i = 0; i < nestedAssetData.length; i++) {
+                uint256 assetAllowance = allowance(owner, spender, nestedAssetData[i]);
```
We should! This was an oversight! Good catch!  Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/acce4f03f8f5a1d914ead1d936fb43555e145480 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+                return 0;
+            }
+        } else if (proxyId == MULTI_ASSET_PROXY_ID) {
+            uint256 lowestAssetAllowance = ~uint256(0);
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/3ae9b5a44c595839cc2ed003a590d0362805aea2 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+            }
+        } else if (proxyId == ERC1155_PROXY_ID) {
+            // solhint-disable-next-line indent
+            (address tokenAddress, , , ) = abi.decode(assetDataBody, (address, uint256[], uint256[], bytes));
```
I changed this to use the new `decodeERC1155AssetData()` function, here: https://github.com/0xProject/0x-monorepo/pull/1779/commits/c4ec07e2883f787345c5eadca291395e399e9fa3

Do you still prefer `LibBytes` here, over re-using this new function?  The new function *does* use `abi.decode`...

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+            return IERC20Token(LibBytes.readAddress(assetData, 16)).allowance(owner, spender);
+        } else if (proxyId == ERC721_PROXY_ID) {
+            (address tokenAddress, uint256 tokenId) = abi.decode(assetDataBody, (address, uint256));
+            if (spender == IERC721Token(tokenAddress).getApproved(tokenId)) {
```
https://github.com/0xProject/0x-monorepo/pull/1779/commits/f93faa067f1953478db41d5eeeb7480e50c2d11d

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+    function batchBalanceOf(address owner, bytes[] memory assetData)
+        public
+        view
+        returns (uint256[] memory amounts)
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/2219036a483a9fd8e0df5862d6d10f45804c3ac4 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+        if (proxyId == ERC20_PROXY_ID) {
+            return IERC20Token(LibBytes.readAddress(assetData, 16)).balanceOf(owner);
+        } else if (proxyId == ERC721_PROXY_ID) {
+            return IERC721Token(LibBytes.readAddress(assetData, 16)).balanceOf(owner);
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/fc10a10ce1ea7420beca57d21dbec74530aa106a .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+        bytes memory assetDataBody = LibBytes.slice(assetData, 4, assetData.length);
+
+        if (proxyId == ERC20_PROXY_ID) {
+            return IERC20Token(LibBytes.readAddress(assetData, 16)).balanceOf(owner);
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/3fbd1133447473e727470ea07571f64afa8039c7 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+    function balanceOf(address owner, bytes memory assetData)
+        public
+        view
+        returns (uint256 amount)
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/2219036a483a9fd8e0df5862d6d10f45804c3ac4 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,250 @@
+            );
+            for (uint256 i = 0; i < nestedAssetData.length; i++) {
+                uint256 assetBalance = balanceOf(owner, nestedAssetData[i]) / assetAmounts[i];
+                if (assetBalance < lowestAssetBalance) {
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/3ae9b5a44c595839cc2ed003a590d0362805aea2 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,228 @@
+            owners[0] = owner;
+            return IERC1155Mintable(tokenAddress).balanceOfBatch(owners, tokenIds)[0];
+        } else if (proxyId == MULTI_ASSET_PROXY_ID) {
+            uint256 lowestAssetBalance = ~uint256(0);
```
Fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/3ae9b5a44c595839cc2ed003a590d0362805aea2 .

feuGeneA commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,228 @@
+        } else if (proxyId == ERC721_PROXY_ID) {
+            return IERC721Token(LibBytes.readAddress(assetData, 16)).balanceOf(owner);
+        } else if (proxyId == ERC1155_PROXY_ID) {
+            (address tokenAddress, uint256[] memory tokenIds, , ) = abi.decode( // solhint-disable-line indent
```
Also fixed in https://github.com/0xProject/0x-monorepo/pull/1779/commits/c890621e3ef967a4d9cf89f700d6eaec13267e86 .

feuGeneA commented:
> I would also prefer to standardize all of the function names to `getBalance`, `getAllowance`, etc. Finally, we should also add `getBalanceAndAllowance` and `getBalancesAndAllowances` functions in addition to the existing batch functions.

@abandeali1 I believe I've addressed all your comments on this PR.  I standardized function names in https://github.com/0xProject/0x-monorepo/pull/1779/commits/77ef4d7489571e84ec27e700ff9a816008da4736 and https://github.com/0xProject/0x-monorepo/pull/1779/commits/ed675c332204cb4656d90822680080e90d81a300 , added those extra balance+allowance methods in https://github.com/0xProject/0x-monorepo/pull/1779/commits/78ef76c152978b99394bb0487b21e1cc5889e95b , and then renamed them to `getBatch...()` instead of `batchGet...()` in https://github.com/0xProject/0x-monorepo/pull/1779/commits/34cf6d79ba92dc19fc11db9340da1e78a98f9677 .

There's one outstanding question to you, [here](https://github.com/0xProject/0x-monorepo/pull/1779#discussion_r277842728).

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,412 @@
+        require(proxyId == ERC1155_PROXY_ID);
+
+        (tokenAddress, tokenIds, tokenValues, callbackData) = abi.decode( // solhint-disable-line indent
+            LibBytes.slice(assetData, 4, assetData.length), (address, uint256[], uint256[], bytes)
```
`slice()` has a lot of overhead. `sliceDestructive()` is probably a better choice here, unless side effect is an issue.

I think you can actually just use assembly to extract these values out by pointing the return variables to values/locations in `assetData`. Ping me if you need help.

dorothy-zbornak replied:
For clarity:

```solidity
assembly {
    // Skip selector and length to get to the first parameter.
    assetData := add(assetData, 36)
    // Read the value of the first parameter.
    tokenAddress := mload(assetData)
    // Point to the next parameter data.
    tokenIds := add(assetData, mload(add(assetData, 32)))
    // Point to the next parameter data.
    tokenvalues := add(assetdata, mload(add(assetdata, 64)))
    // Point to the next parameter data.
    callbackData := add(assetData, add(mload(assetData), 96)))
}
```

abandeali1 replied:
I'd prefer to minimize the use of assembly and prioritize readability for this contract, since it will probably just be used for calls where efficiency isn't a consideration. I don't mind using `sliceDestructive` here though.

abandeali1 replied:
Actually, thinking about it more, I could see this being inherited and used internally within other contracts pretty frequently. In that case I'd rather not destroy memory, but efficiency does become more of a factor.

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,412 @@
+        require(proxyId == MULTI_ASSET_PROXY_ID);
+
+        // solhint-disable-next-line indent
+        (amounts, nestedAssetData) = abi.decode(LibBytes.slice(assetData, 4, assetData.length), (uint256[], bytes[]));
```
Can optimize with `sliceDestructive()`/assembly here too.

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,420 @@
+            }
+            return lowestAssetBalance;
+        } else {
+            revert("Unsupported proxy identifier");
```
Sorry, missed this first round. Our convention is to revert with `"ALL_CAPS_SNAKE_CASE"` error messages.

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/libs/LibAssetData.sol:
```diff
@@ -0,0 +1,420 @@
+            }
+            return lowestAssetAllowance;
+        } else {
+            revert("Unsupported proxy identifier");
```
Snake-case here, too.
